llvm/tools/llvm-cas-dump/MCCASPrinter.cpp

```diff
@@ -50,24 +55,43 @@ Error MCCASPrinter::printMCObject(ObjectRef CASObj) {
   Expected<MCObjectProxy> MCObj = MCSchema.get(CASObj);
   if (!MCObj)
     return MCObj.takeError();
-  return printMCObject(*MCObj);
+  return printMCObject(*MCObj, DWARFCtx);
 }
 
-Error MCCASPrinter::printMCObject(MCObjectProxy MCObj) {
+Error MCCASPrinter::printMCObject(MCObjectProxy MCObj, DWARFContext *DWARFCtx) {
+  // Initialize DWARFObj and scan and make a first pass through the sections.
+  std::unique_ptr<DWARFContext> DWARFContextHolder;
+  if (Options.DwarfDump && !DWARFCtx) {
+    auto DWARFObj = std::make_unique<CASDWARFObject>(MCObj.getSchema());
+    if (Error err = DWARFObj->discoverDwarfSections(MCObj))
+      return err;
+    DWARFContextHolder = std::make_unique<DWARFContext>(std::move(DWARFObj));
+    DWARFCtx = DWARFContextHolder.get();
+  }
+
   // If only debug sections were requested, skip non-debug sections.
   if (Options.DwarfSectionsOnly && SectionRef::Cast(MCObj) &&
       !isDwarfSection(MCObj))
     return Error::success();
 
+  // Print CAS Id.
   OS.indent(Indent);
   OS << formatv("{0, -15} {1} \n", MCObj.getKindString(),
                 MCSchema.CAS.getID(MCObj));
 
-  return printSimpleNested(MCObj);
+  // Dwarfdump.
+  if (DWARFCtx) {
+    auto *DWARFObj =
```

Comment:
It think we could avoid hardcoding the Indent constant by:
```cpp
IndentGuard Guard(Indent);
```

And then later we can remove the `+2`